File: windows/src/GoalMaker.Core/Sync/SyncedTable.cs
namespace GoalMaker.Core.Sync;

/// <summary>A table both apps replicate, with its columns in contract order.</summary>
public sealed record SyncedTable(string Name, IReadOnlyList<SyncedColumn> Columns)
{
    /// <summary>The columns the server needs a value in, so a new row has to carry all of them.</summary>
    public IReadOnlyList<string> Required { get; } =
        [.. Columns.Where(column => column.Required).Select(column => column.Name)];

    public const string Id = "id";
    public const string OwnerId = "owner_id";
    public const string CreatedAt = "created_at";
    public const string UpdatedAt = "updated_at";
    public const string DeletedAt = "deleted_at";
}

File: android/app/src/main/kotlin/com/goalmaker/app/domain/sync/SyncedTable.kt
package com.goalmaker.app.domain.sync

/** A table both apps replicate, with its columns in contract order. */
data class SyncedTable(
    val name: String,
    val columns: List<SyncedColumn>,
) {
    /** The columns the server needs a value in, so a new row has to carry all of them. */
    val required: List<String> = columns.filter(SyncedColumn::required).map(SyncedColumn::name)

    companion object {
        const val ID = "id"
        const val OWNER_ID = "owner_id"
        const val CREATED_AT = "created_at"
        const val UPDATED_AT = "updated_at"
        const val DELETED_AT = "deleted_at"
    }
}
